django/contrib/contenttypes/fields.py

```diff
@@ -505,6 +505,17 @@ def related_manager_cls(self):
             self.rel,
         )
 
+    @cached_property
+    def related_manager_cache_key(self):
+        # By default, GenericRel instances will be marked as hidden (their
+        # accessor name being `+`), which would cause multiple GenericRelations
+        # declared on a single model to collide, so use the remote field's name.
+        if self.rel.hidden:
+            return self.field.get_cache_name()
+        # If the field isn't hidden (e.g: having been made public via user
+        # subclassing), use the related_query_name value
+        return self.rel.name
```

Comment:
On reflection, having slept on it, I'm not sure _why_ I think this is necessary. I _think_ I can maybe just use the `self.field.get_cache_name()` always and this is just over-complicating it.

It _does_ get exercised by tests though, via `test_inherited_models_content_type` and `test_generic_rel` and `test_queries_content_type_restriction` for example. Hmmm.

Reply:
Introduced a test (`test_gfk_related_manager_no_overlap_when_not_hidden`) + model (`ConfusedBook`) to demonstrate the slightly convoluted circumstances in which using `self.rel.name` would've been wrong; as such I'm now only using `self.field.get_cache_name()` to make that test pass. Which means (somewhat sensibly) that the m2m one is the only branching version.